builtins/amp-img.js

```diff
@@ -45,11 +46,39 @@ const ATTRIBUTES_TO_PROPAGATE = [
 ];
 
 export class AmpImg extends BaseElement {
+  /** @override @nocollapse */
+  static V2() {
+    return V2_IMG_VIDEO;
+  }
+
   /** @override @nocollapse */
   static prerenderAllowed() {
     return true;
   }
 
+  /** @override @nocollapse */
+  static getPreconnects(element) {
+    const src = element.getAttribute('src');
+    if (src) {
+      return [src];
+    }
+
+    // NOTE(@wassgha): since parseSrcset is computationally expensive and can
+    // not be inside the `buildCallback`, we went with preconnecting to the
+    // `src` url if it exists or the first srcset url.
+    const srcset = element.getAttribute('srcset');
+    if (srcset) {
+      // We try to find the first url in the srcset
+      const srcseturl = /\S+/.exec(srcset);
+      // Connect to the first url if it exists
+      if (srcseturl) {
+        return [srcseturl[0]];
+      }
+    }
+
+    return null;
```

Comment:
I'm currently supporting either. The base class just returns `null` to reduce allocations since most of subclasses do not have any preconnects. I think that's justified since the whole point of this structure is to reduce number of allocations until later.